### ISSUE TITLE
Question about deformable roi pooling.

### ISSUE THREAD

**Issue by youngfly11** [jwyang/faster-rcnn.pytorch#237]
## rois align
As we see in rois align code, the function just return the  gradient of feature maps. The gradient of rois is not returned. Because we have no request of rois' gradients.

But in the deformable rois pooling, the rois are separated into $K*K$ bins. we can add some offset to each bins。So I need to require the gradient of rois. I have try write the backward function by myself.
This is the code in rois align functions:
```python
 def backward(self, grad_output):
        assert(self.feature_size is not None and grad_output.is_cuda)

        batch_size, num_channels, data_height, data_width = self.feature_size
        grad_input = self.rois.new(batch_size, num_channels, data_height,
                                   data_width).zero_()

        offset_x_grad = self.rois.new(self.num_rois, num_channels, self.aligned_height, self.aligned_width).zero_()
        offset_y_grad = self.rois.new(self.num_rois, num_channels, self.aligned_height, self.aligned_width).zero_()

        roi_plus.roi_plus_backward_cuda(self.aligned_height,
                                        self.aligned_width,
                                        self.spatial_scale, grad_output,
                                        self.rois, grad_input,
                                        offset_x_grad, offset_y_grad,
                                        self.feature)

        offset_x_grad = torch.sum(torch.sum(torch.sum(offset_x_grad, -1), -1), -1) ## 3
        offset_y_grad = torch.sum(torch.sum(torch.sum(offset_y_grad, -1), -1), -1) ## 3

        grad_rois = self.rois.new(self.num_rois, 5).zero_()
        for idx in range(self.num_rois):

            grad_rois[idx, 1] = grad_rois[idx, 1] + offset_x_grad[idx]
            grad_rois[idx, 3] = grad_rois[idx, 3] + offset_x_grad[idx]
            grad_rois[idx, 2] = grad_rois[idx, 2] + offset_y_grad[idx]
            grad_rois[idx, 4] = grad_rois[idx, 4] + offset_y_grad[idx]

        return grad_input, grad_rois
```
This is the backward code in rois_align_kernel.cu
```c
 // bilinear interpolation
            if (!(h < 0 || h >= height || w < 0 || w >= width)) {
                float h_ratio = h - (float)(hstart);
                float w_ratio = w - (float)(wstart);
                int upleft = img_start + (c * height + hstart) * width + wstart;
                int upright = upleft + 1;
                int downleft = upleft + width;
                int downright = downleft + 1;

                atomicAdd(bottom_diff + upleft, top_diff[index] * (1. - h_ratio) * (1 - w_ratio));
                atomicAdd(bottom_diff + upright, top_diff[index] * (1. - h_ratio) * w_ratio);
                atomicAdd(bottom_diff + downleft, top_diff[index] * h_ratio * (1 - w_ratio));
                atomicAdd(bottom_diff + downright, top_diff[index] * h_ratio * w_ratio);


                // calculate offset grad for x and y
                atomicAdd(offset_x_diff + index, top_diff[index] * ( - feature[upleft] * (1. - h_ratio) +
                                                                     feature[upright] * (1. - h_ratio) +
                                                                   - feature[downleft] * h_ratio + 
                                                                     feature[downright] * h_ratio ));
                                                                     
                atomicAdd(offset_y_diff + index, top_diff[index] * ( - feature[upleft] * (1. - w_ratio) +
                                                                   - feature[upright] * w_ratio +
                                                                     feature[downleft] * (1. - w_ratio) + 
                                                                     feature[downright] * w_ratio ));
```
I am not very confident about the backpropagation grdient of rois. So I need someone to help give some advices about gradients of rois or help me correct the mistakes in my codes. 
Thank you.
